lms/djangoapps/completion/models.py

```diff
@@ -105,6 +127,91 @@ def submit_completion(self, user, course_key, block_key, completion):
             )
         return obj, isnew
 
+    def submit_batch_completion(self, user, course_key, blocks):
+        """
+        Performs a batch insertion of completion objects.
+
+        Parameters:
+            * user (django.contrib.auth.models.User): The user for whom the
+              completions are being submitted.
+            * course_key (opaque_keys.edx.keys.CourseKey): The course in
+              which the submitted blocks are found.
+            * blocks: A list tuple of UsageKey to float completion values.
+              (float in range [0.0, 1.0]): The fractional completion
+              value of the block (0.0 = incomplete, 1.0 = complete).
+
+        Return Value:
+            Dict of (BlockCompletion, bool): A dictionary with a
+            BlockCompletion object key and a value of bool. The boolean value
+            indicates whether the object was newly created by this call.
+
+        Raises:
+
+            ValueError:
+                If the wrong type is passed for one of the parameters.
+
+            django.core.exceptions.ValidationError:
+                If a float is passed that is not between 0.0 and 1.0.
+
+            django.db.DatabaseError:
+                If there was a problem getting, creating, or updating the
+                BlockCompletion record in the database.
+
+                This will also be a more specific error, as described here:
+                https://docs.djangoproject.com/en/1.11/ref/exceptions/#database-exceptions.
+                IntegrityError and OperationalError are relatively common
+                subclasses.
+        """
```

Comment:
Django provides a lovely [transaction.atomic](https://docs.djangoproject.com/en/1.11/topics/db/transactions/#controlling-transactions-explicitly
) decorator/context manager which forces the transactions within a block to be atomic.


I think you can use this to wrap calls to the existing `submit_completion` method and get them done in one transaction, which means most of this code can be removed.  e.g.

```python
with transaction.atomic():
    for block, completion in blocks:
        self.submit_completion(user, course_key, block, completion):
```

Reply:
I ran some tests, and you were right!